aidial_adapter_openai/gpt4_multi_modal/chat_completion.py

```diff
@@ -180,26 +188,31 @@ async def chat_completion(
 
     api_url = f"{upstream_endpoint}?api-version={api_version}"
 
-    result = await transform_messages(file_storage, messages)
+    result = await transform_messages(file_storage, messages, tokenizer)
 
-    if isinstance(result, str):
+    if result.error_message:
         logger.error(f"Failed to prepare request: {result}")
 
         chunk = create_stage_chunk("Usage", USAGE, is_stream)
 
-        exc = InvalidRequestError(message=result, display_message=result)
+        exc = InvalidRequestError(
+            message=result.error_message, display_message=result
+        )
 
         return create_response_from_chunk(chunk, exc, is_stream)
 
-    new_messages, prompt_image_tokens = result
+    max_tokens = request.get("max_tokens") or default_max_tokens
 
-    prompt_text_tokens = tokenizer.calculate_prompt_tokens(messages)
-    estimated_prompt_tokens = prompt_text_tokens + prompt_image_tokens
+    if result.transformations is None:
+        raise InternalServerError("Transformations cannot be null")
+    truncation_result = truncate_transformed_messages(
+        result.transformations, max_tokens
+    )
 
     request = {
         **request,
-        "max_tokens": request.get("max_tokens") or default_max_tokens,
-        "messages": new_messages,
+        "max_tokens": max_tokens,
+        "messages": truncation_result.messages,
     }
 
```

Comment:
Destructure to reduce verbosity below.

`estimated_prompt_tokens, discarded_messages = truncation_result.overall_token, truncation_result.discarded_messages`

Reply:
Not actual anymore, function has changed drastically